tests/Http/AuthenticatorTest.php

```diff
@@ -43,4 +50,189 @@ public function testSetsAndProvidesCredentials()
+        $authenticator->fetchAccessAndRefreshToken();
+
+        $this->assertTrue($callback->executed);
+        $this->assertEquals($accessToken, $callback->accessToken);
+        $this->assertEquals($expiresIn, $callback->expiresIn);
+        $this->assertEquals($refreshToken, $callback->refreshToken);
+    }
+
+    /**
+     * @dataProvider autoRefreshAccessTokenProvider
+     */
+    public function testIdentifiesWhenToAutoRefreshToken(array $args)
+    {
+        // setting default values to keep phpstan happy
+        $tokenType = null;
+        $tokenCallback = function () {};
+        $shouldAutoRefreshAccessToken = null;
+
+        /*
+         * @var string $tokenType
+         * @var null|Closure $tokenCallback
+         * @var bool $shouldAutoRefreshAccessToken
+         */
+        extract($args);
+
+        $authType = \Mockery::mock(Auth::class);
+        $authType->shouldReceive('getType')->andReturn($tokenType);
+
+        $authenticator = new Authenticator($this->client, $authType);
+        $authenticator->callbackWhenTokenRefreshed($tokenCallback);
+
+        $this->assertEquals(
+            $shouldAutoRefreshAccessToken,
+            $authenticator->shouldAutoRefreshAccessToken()
+        );
+    }
+
+    public function autoRefreshAccessTokenProvider(): \Generator
+    {
+        $typesThatCanRefreshTokens = [
+            ClientCredentials::TYPE,
+            RefreshCredentials::TYPE,
+            CodeCredentials::TYPE,
+        ];
+        foreach ($typesThatCanRefreshTokens as $type) {
+            yield ['Should not refresh because there is no callback' => [
+                'tokenType' => $type,
+                'tokenCallback' => null,
+                'shouldAutoRefreshAccessToken' => false,
+            ]];
+            yield ['Should refresh when callback and correct auth type' => [
+                'tokenType' => $type,
+                'tokenCallback' => function () {
+                    return true;
+                },
+                'shouldAutoRefreshAccessToken' => true,
+            ]];
+            yield [[
+                'tokenType' => $type,
+                'tokenCallback' => function () {
+                    return true;
+                },
+                'shouldAutoRefreshAccessToken' => true,
+            ]];
+        }
+
+        $typesThatCannnotRefreshTokens = [
+            NullCredentials::TYPE,
+        ];
+        foreach ($typesThatCannnotRefreshTokens as $type) {
+            yield ['Should not refresh because there is no callback' => [
+                'tokenType' => $type,
+                'tokenCallback' => null,
+                'shouldAutoRefreshAccessToken' => false,
+            ]];
+            yield ['Should not refresh when there is a callback' => [
+                'tokenType' => $type,
+                'tokenCallback' => function () {
+                    return true;
+                },
+                'shouldAutoRefreshAccessToken' => false,
+            ]];
+        }
```

Comment:
Does it make sense for this to be a separate test?

Reply:
It _could_ be, but it would largely be the same as `testIdentifiesWhenToAutoRefreshToken()`, so I left it here to avoid that duplication.